akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -432,594 +525,144 @@ private[akka] class Shard(
     }
   }
 
-  // EntityStopped handler
-  def passivateCompleted(event: EntityStopped): Unit = {
-    val hasBufferedMessages = messageBuffers.getOrEmpty(event.entityId).nonEmpty
-    state = state.copy(state.entities - event.entityId)
+  // After entity stopped
+  def passivateCompleted(entityId: EntityId): Unit = {
+    val hasBufferedMessages = messageBuffers.getOrEmpty(entityId).nonEmpty
+    entityIds = entityIds - entityId
     if (hasBufferedMessages) {
-      log.debug(
-        "Entity stopped after passivation [{}], but will be started again due to buffered messages.",
-        event.entityId)
-      processChange(EntityStarted(event.entityId))(sendMsgBuffer)
+      log.debug("Entity stopped after passivation [{}], but will be started again due to buffered messages.", entityId)
+      waitForAsyncWrite("remove (passivated)", entityId, rememberEntitiesStore.addEntity(entityId))(sendMsgBuffer)
     } else {
-      log.debug("Entity stopped after passivation [{}]", event.entityId)
-      messageBuffers.remove(event.entityId)
+      log.debug("Entity stopped after passivation [{}]", entityId)
+      dropBufferFor(entityId)
     }
 
   }
 
-  // EntityStarted handler
-  def sendMsgBuffer(event: EntityStarted): Unit = {
-    //Get the buffered messages and remove the buffer
-    val messages = messageBuffers.getOrEmpty(event.entityId)
-    messageBuffers.remove(event.entityId)
-
-    if (messages.nonEmpty) {
-      log.debug("Sending message buffer for entity [{}] ([{}] messages)", event.entityId, messages.size)
-      getOrCreateEntity(event.entityId)
-      //Now there is no deliveryBuffer we can try to redeliver
-      // and as the child exists, the message will be directly forwarded
-      messages.foreach {
-        case (msg, snd) => deliverMessage(msg, snd)
-      }
-    }
-  }
-
-  def deliverMessage(msg: Any, snd: ActorRef): Unit = {
+  /**
+   * @param entityIdWaitingForWrite an id for an remember entity write in progress, if non empty messages for that id
+   *                                will be buffered
+   */
+  private def deliverMessage(msg: Any, snd: ActorRef, entityIdWaitingForWrite: OptionVal[EntityId]): Unit = {
```

Comment:
This is difficult, but also the area we want to explore performance optimizations for so more changes will probably come. 

Just some early thoughts for later. It can currently not start more than one entity at a time. That's how it was before also, but we could maybe have a store that can store for different entity ids in parallel. A single EventSourcedStore can't but the we could use more than one and hash/group on the entityId in similar way as we do for ddata.

Reply:
One idea I had for improvement here would be to replace several lookups (messageBuffers, entityWaitingForWrite, refById, entityIds) with a single lookup, like a `statePerEntity: Map[EntityId, State]` where `State` could be something like `Idle(ref)`, `Passivating(ref)`, `WritInProgress(ref)` and perhaps a default `NotRunning` avoiding wrapping/unwrapping an `Option`.

Not so much for performance as for making it easier to follow.

Reply:
Why can't we do it with a single event sourced store given we have one per shard? If the API took a AddEntities rather than AddEntity we could batch together multiple and persist them with a single persist.

At the moment we have the messages in a stash but we could have them in actor state and then when a write is complete to the remembered entities store send the rest as a batch?

I think this is a tradeoff between the simplicity of stashing and cost of doing one thing at a time.